{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create Population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# n: number of actors\n",
    "# g: number of genes per actor\n",
    "def createRandomPopulation(n, g):\n",
    "    '''Returns array of indices array[actor][geneSet][gene]'''\n",
    "\n",
    "    actors = []\n",
    "    genotype = []\n",
    "    for i in range(0,n):\n",
    "        genotype = []\n",
    "        for i in range (0, 2):\n",
    "            genes = []\n",
    "            for i in range(0, g):\n",
    "                genes.append(random.choice([0,1]))\n",
    "            \n",
    "            genotype.append(genes)\n",
    "        actors.append(genotype)\n",
    "        \n",
    "    \n",
    "    return actors\n",
    "\n",
    "createRandomPopulation(10, 10)\n",
    "\n",
    "\n",
    "        \n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Evolution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fitness: Probability of survival based on amount of recessive phenotype - meaning genes are 0 and 0.\n",
    "def Fitness(actor, k):\n",
    "    r = 0\n",
    "    for i in actor[0]:\n",
    "        #if phenotype is recessive\n",
    "        if (actor[0][i] == [0] and actor[1][i] == [0]):\n",
    "            r += 1\n",
    "    #Linear correlation between survival chance and recessive phenotype\n",
    "    return (1 - k*(r / len(actor[0])))\n",
    "\n",
    "\n",
    "# Split population in two\n",
    "def split_list(a_list):\n",
    "    half = len(a_list)//2\n",
    "    return a_list[:half], a_list[half:]\n",
    "\n",
    "\n",
    "def removeActors(actors, k):\n",
    "    for actor in actors:\n",
    "        #Simulating Actor Fitness probability\n",
    "        if Fitness(actor, k)*100 < random.randint(1, 100):\n",
    "            actors.remove(actor)\n",
    "\n",
    "# Mitosis Simulation\n",
    "def newGenActorCreation(actors):\n",
    "    #split the array randomly. If odd, kill (or ignore) last one.\n",
    "    a, b = split_list(actors)\n",
    "    # print(a)\n",
    "    # print(b)\n",
    "    genesPassedFromActorsA = []\n",
    "    genesPassedFromActorsB = []\n",
    "    genesPassedCombined = []\n",
    "    # This array is determining the amount of gametes each member of sets A and B produce. \n",
    "    # The number of gametes from A and B have to be the same to produce a new generation.\n",
    "    OffspringAmount = []\n",
    "    actor_indice_a = 0\n",
    "    actor_indice_b = 0\n",
    "    for i in range(0, len(b)):\n",
    "        r = random.randrange(1, 100)\n",
    "        if r > 50:\n",
    "            OffspringAmount.append(2)\n",
    "        elif r > 30:\n",
    "            OffspringAmount.append(1)\n",
    "        else:\n",
    "            OffspringAmount.append(3)\n",
    "    # print(OffspringAmount)    \n",
    "    \n",
    "    for actor in a:\n",
    "        \n",
    "        for j in range(OffspringAmount[actor_indice_a]):\n",
    "            genesPassed = []\n",
    "            for k in range(0, len(actor[0])):\n",
    "                genesPassed.append(random.choice([actor[0][j], actor[1][j]]))\n",
    "            genesPassedFromActorsA.append(genesPassed)\n",
    "        # print(OffspringAmount[actor_indice_a], actor_indice_a)\n",
    "        actor_indice_a += 1\n",
    "        \n",
    "    # print(genesPassedFromActorsA)\n",
    "\n",
    "\n",
    "\n",
    "    for actor in b:\n",
    "        \n",
    "        for j in range(OffspringAmount[actor_indice_b]):\n",
    "            genesPassed = []\n",
    "            for k in range(0, len(actor[0])):\n",
    "                genesPassed.append(random.choice([actor[0][j], actor[1][j]]))\n",
    "            genesPassedFromActorsB.append(genesPassed)\n",
    "        actor_indice_b += 1\n",
    "        \n",
    "    # print(genesPassedFromActorsB)\n",
    "\n",
    "\n",
    "    for i in range (0, len(genesPassedFromActorsA)):\n",
    "        genesPassedCombined.append([genesPassedFromActorsA[i], genesPassedFromActorsB[i]])\n",
    "    return genesPassedCombined\n",
    "\n",
    "# p: probability of mutation\n",
    "def mutate(actors, p):\n",
    "    a = []\n",
    "    for actor in actors:\n",
    "        b = []\n",
    "        for geneset in actor:\n",
    "            c = []\n",
    "            for gene in geneset:\n",
    "                if random.randint(1, 100) < p*100:\n",
    "                    if gene == 1:\n",
    "                        c.append(0)\n",
    "                    else: \n",
    "                        c.append(1)\n",
    "                else:\n",
    "                    c.append(gene)\n",
    "            b.append(c)\n",
    "        a.append(b)\n",
    "    return a\n",
    "                \n",
    "\n",
    "# print(f\" New Gen with mutations: {mutate(test, 0.3)}\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "        \n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running Simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.ylabel('Ratio of Recessive Alleles')\n",
    "plt.xlabel('Generation')\n",
    "\n",
    "InitialActors = createRandomPopulation(50, 50)\n",
    "# Actors: initial population, p: probability of mutation, t: number of generation simulations runs, k: fitness probabilty constant (higher means more punishment for recessive phenotype)\n",
    "def runSimulation(actors, p, t, k):\n",
    "    a = actors\n",
    "    x = []\n",
    "    y = []\n",
    "    for i in range(0, t):\n",
    "        recessive_allele = 0\n",
    "        removeActors(a, k)\n",
    "        total = 0\n",
    "\n",
    "        a = newGenActorCreation(a)\n",
    "\n",
    "        mutate(a, p)\n",
    "\n",
    "        # print(f\"Actors in Gen {i+1}: {a}\")\n",
    "\n",
    "        for actors in a:\n",
    "            for genesets in actors:\n",
    "                for allele in genesets:\n",
    "                    if allele == 0:\n",
    "                        recessive_allele += 1\n",
    "                    total += 1\n",
    "\n",
    "\n",
    "        if total == 0:\n",
    "            ratio = 0\n",
    "        else:\n",
    "            ratio = recessive_allele/total\n",
    "        print(f\"Gen: {i+1}, Total Alleles: {total}, Recessive alleles: {recessive_allele}, Ratio: {ratio}\")\n",
    "        x.append(i+1)\n",
    "        y.append(ratio*100)\n",
    "    print(x)\n",
    "    print(y)\n",
    "    plt.plot(y)\n",
    "    plt.show()\n",
    "\n",
    "        \n",
    "runSimulation(InitialActors, 0.3, 100, 3)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "compbio",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "2faf86e960bc6613e78dc13eb0f02b68a42c32bcd2ce585b9de86e4da29ebc1b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
